File: go/s2iam/jwt.go
package s2iam

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/memsql/errors"
	"github.com/singlestore-labs/singlestore-auth-iam/go/s2iam/models"
)

const (
	// defaultServer is the default authentication server endpoint
	defaultServer = "https://authsvc.singlestore.com/auth/iam/:jwtType"
)

// JWTOptions are used to configure how to get JWTs
type JWTOption interface {
	applyJWTOption(*jwtOptions)
}

// Implementation struct for JWT options
type jwtOption func(*jwtOptions)

func (o jwtOption) applyJWTOption(opts *jwtOptions) {
	o(opts)
}

// jwtOptions holds the options for the getJWT function
type jwtOptions struct {
	detectProviderOptions
	JWTType              JWTType
	WorkspaceGroupID     string
	ServerURL            string
	Provider             models.CloudProviderClient
	AdditionalParams     map[string]string
	AssumeRoleIdentifier string
}

// WithServerURL sets the authentication server URL
func WithServerURL(serverURL string) JWTOption {
	return jwtOption(func(o *jwtOptions) {
		o.ServerURL = serverURL
	})
}

// WithProvider sets a specific cloud provider client to use
func WithProvider(provider models.CloudProviderClient) JWTOption {
	return jwtOption(func(o *jwtOptions) {
		o.Provider = provider
	})
}

// WithGCPAudience sets the GCP audience for identity token requests
func WithGCPAudience(audience string) JWTOption {
	return jwtOption(func(o *jwtOptions) {
		o.AdditionalParams["audience"] = audience
	})
}

// WithAssumeRole sets the role identifier to assume (if there is one)
func WithAssumeRole(roleIdentifier string) JWTOption {
	return jwtOption(func(o *jwtOptions) {
		o.AssumeRoleIdentifier = roleIdentifier
	})
}

// processJWTOptions processes JWT options and extracts provider options
func processJWTOptions(jwtOpts jwtOptions, opts ...JWTOption) jwtOptions {
	if jwtOpts.AdditionalParams == nil {
		jwtOpts.AdditionalParams = make(map[string]string)
	}

	//nolint:staticcheck // QF1008: could remove embedded field "detectProviderOptions" from selector
	if jwtOpts.detectProviderOptions.timeout == 0 {
		//nolint:staticcheck // QF1008: could remove embedded field "detectProviderOptions" from selector
		jwtOpts.detectProviderOptions.timeout = defaultTimeout
	}

	for _, opt := range opts {
		// Apply to both option types
		opt.applyJWTOption(&jwtOpts)
	}

	return jwtOpts
}

// getJWT retrieves a JWT from the authentication server using cloud provider identity
func getJWT(ctx context.Context, defaultOpts jwtOptions, opts []JWTOption) (string, error) {
	jwtOpts := processJWTOptions(defaultOpts, opts...)

	if jwtOpts.ServerURL == "" {
		return "", errors.New("server URL is required")
	}

	// Auto-detect provider if not specified
	if jwtOpts.Provider == nil {
		var err error
		jwtOpts.Provider, err = detectProviderImpl(ctx, jwtOpts.detectProviderOptions)
		if err != nil {
			return "", errors.Errorf("failed to detect cloud provider: %w", err)
		}
	}

	// Create provider with assumed role if needed
	provider := jwtOpts.Provider
	if jwtOpts.AssumeRoleIdentifier != "" {
		provider = provider.AssumeRole(jwtOpts.AssumeRoleIdentifier)
	}

	identityHeaders, identity, err := provider.GetIdentityHeaders(ctx, jwtOpts.AdditionalParams)
	if err != nil {
		return "", errors.Errorf("failed to get identity headers: %w", err)
	}

	// Construct the URL
	targetURL := jwtOpts.ServerURL
	targetURL = strings.ReplaceAll(targetURL, ":cloudProvider", string(identity.Provider))
	targetURL = strings.ReplaceAll(targetURL, ":jwtType", string(jwtOpts.JWTType))

	uri, err := url.Parse(targetURL)
	if err != nil {
		return "", errors.Errorf("invalid server URL: %w", err)
	}

	// Add query parameters
	q := uri.Query()
	if jwtOpts.JWTType == DatabaseAccessJWT && jwtOpts.WorkspaceGroupID != "" {
		q.Add("workspaceGroupID", jwtOpts.WorkspaceGroupID)
	}
	uri.RawQuery = q.Encode()

	// Create request
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, uri.String(), nil)
	if err != nil {
		return "", errors.Errorf("error creating request: %w", err)
	}

	// Add identity headers
	for key, value := range identityHeaders {
		req.Header.Set(key, value)
	}

	// Send request
	httpClient := &http.Client{
		Timeout: 10 * time.Second, // Default HTTP client timeout
	}
	resp, err := httpClient.Do(req)
	if err != nil {
		return "", errors.Errorf("error calling authentication server: %w", err)
	}
	defer func() {
		_ = resp.Body.Close()
	}()

	// Process response
	bodyBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", errors.Errorf("error reading response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return "", errors.Errorf("authentication server returned status %d: %s", resp.StatusCode, string(bodyBytes))
	}

	var response struct {
		JWT string `json:"jwt"`
	}
	if err := json.Unmarshal(bodyBytes, &response); err != nil {
		return "", errors.Errorf("cannot parse response: %w", err)
	}

	if response.JWT == "" {
		return "", errors.New("received empty JWT from server")
	}

	return response.JWT, nil
}

// GetDatabaseJWT retrieves a database JWT from the authentication server
func GetDatabaseJWT(ctx context.Context, workspaceGroupID string, opts ...JWTOption) (string, error) {
	if workspaceGroupID == "" {
		return "", errors.New("workspaceGroupID is required for database JWT")
	}

	return getJWT(ctx, jwtOptions{
		JWTType:          DatabaseAccessJWT,
		WorkspaceGroupID: workspaceGroupID,
		ServerURL:        defaultServer,
	}, opts)
}

// GetAPIJWT retrieves an API JWT from the authentication server
func GetAPIJWT(ctx context.Context, opts ...JWTOption) (string, error) {
	return getJWT(ctx, jwtOptions{
		JWTType:   APIGatewayAccessJWT,
		ServerURL: defaultServer,
	}, opts)
}

File: python/src/s2iam/jwt.py
"""
JWT functionality for SingleStore authentication.
"""

import asyncio
import os
import socket
import time
from typing import Any, Optional

import aiohttp

from .models import (
    CloudProviderClient,
    JWTType,
    Logger,
)

DEFAULT_SERVER_URL = "https://authsvc.singlestore.com/auth/iam/{jwt_type}"


async def get_jwt(
    jwt_type: JWTType,
    workspace_group_id: Optional[str] = None,
    server_url: Optional[str] = None,
    provider: Optional[CloudProviderClient] = None,
    additional_params: Optional[dict[str, str]] = None,
    assume_role_identifier: Optional[str] = None,
    timeout: float = 10.0,
    logger: Optional[Logger] = None,
    **kwargs: Any,
) -> str:
    """
    Get a JWT from SingleStore's authentication service.

    This function attempts to obtain a JWT from SingleStore's authentication service
    using the detected cloud provider's identity.

    Args:
        jwt_type (JWTType): The type of JWT to request (database, api)
        workspace_group_id (Optional[str]): Workspace group ID to scope the JWT to.
            Only used for database JWTs. When None, the JWT may have broader access.
        timeout (float): Timeout in seconds for the request
        server_url (Optional[str]): Override the default server URL
        logger (Optional[Logger]): Logger instance for debug output

    Returns:
        str: JWT string

    Raises:
        NoCloudProviderDetectedError: If no cloud provider is detected
        Exception: If JWT acquisition fails
    """
    # Detect provider if not provided
    if provider is None:
        # Import here to avoid circular import
        from .api import detect_provider

        provider = await detect_provider(logger=logger, **kwargs)

    # Assume role if requested
    if assume_role_identifier:
        provider = provider.assume_role(assume_role_identifier)

    # Get identity headers
    headers, identity = await provider.get_identity_headers(additional_params)

    # Prepare server URL, allow override via environment variable
    env_server_url = os.environ.get("S2IAM_JWT_SERVER_URL")
    if server_url is None:
        if env_server_url:
            server_url = env_server_url
        else:
            server_url = DEFAULT_SERVER_URL.format(jwt_type=jwt_type.value)

    # Log request if logger available (aligned with Go: no JSON body sent, identity conveyed via headers only)
    if logger:
        logger.log(
            "Requesting JWT (type="
            f"{jwt_type.value}) from {server_url} provider={identity.provider.value} "
            f"workspace_group_id={workspace_group_id or '<none>'}"
        )

    connector = None
    if os.environ.get("S2IAM_FORCE_IPV4") == "1":  # optional mitigation for IPv6 stalls
        connector = aiohttp.TCPConnector(family=socket.AF_INET)

    start = time.monotonic()

    # Lightweight phase tracing (only adds small overhead). Captures which phase we reached
    # so timeout/connect errors are more actionable without deep instrumentation.
    phase: str = "init"
    phases: list[str] = []

    trace = aiohttp.TraceConfig()

    from aiohttp import ClientSession
    from aiohttp.tracing import (
        TraceConnectionCreateEndParams,
        TraceConnectionCreateStartParams,
        TraceRequestEndParams,
        TraceRequestStartParams,
    )

    @trace.on_request_start.append
    async def _on_request_start(
        session: ClientSession,
        ctx: Any,
        params: TraceRequestStartParams,
    ) -> None:  # type: ignore[unused-ignore]
        nonlocal phase
        phase = "request_start"
        phases.append(phase)

    @trace.on_connection_create_start.append
    async def _on_conn_start(
        session: ClientSession,
        ctx: Any,
        params: TraceConnectionCreateStartParams,
    ) -> None:  # type: ignore[unused-ignore]
        nonlocal phase
        phase = "connect_start"
        phases.append(phase)

    @trace.on_connection_create_end.append
    async def _on_conn_end(
        session: ClientSession,
        ctx: Any,
        params: TraceConnectionCreateEndParams,
    ) -> None:  # type: ignore[unused-ignore]
        nonlocal phase
        phase = "connect_end"
        phases.append(phase)

    @trace.on_request_end.append
    async def _on_request_end(
        session: ClientSession,
        ctx: Any,
        params: TraceRequestEndParams,
    ) -> None:  # type: ignore[unused-ignore]
        nonlocal phase
        phase = "request_end"
        phases.append(phase)

    try:
        async with aiohttp.ClientSession(
            timeout=aiohttp.ClientTimeout(total=timeout),
            connector=connector,
            trace_configs=[trace],
        ) as session:
            async with session.post(
                server_url,
                headers={
                    **headers,
                    "Content-Type": "application/json",
                },
                # No body (parity with Go implementation)
                data=None,
            ) as response:
                elapsed = (time.monotonic() - start) * 1000.0
                if response.status == 200:
                    # Expect JSON with {"jwt": "..."}
                    response_data = await response.json()
                    jwt_value = response_data.get("jwt")
                    if not isinstance(jwt_value, str) or not jwt_value:
                        raise Exception("No JWT in response")
                    if logger and os.environ.get("S2IAM_DEBUGGING") == "true":
                        logger.log(f"JWT obtained in {elapsed:.1f}ms")
                    return jwt_value
                error_text = await response.text()
                raise Exception(
                    "JWT request failed status="
                    f"{response.status} elapsed_ms={elapsed:.1f} phase={phase} body_snip={error_text[:160]!r}"
                )
    except asyncio.TimeoutError:
        raise Exception(
            "JWT request timeout after "
            f"{timeout}s (elapsed_ms={(time.monotonic()-start)*1000.0:.1f} last_phase={phase} phases={phases})"
        )


# Legacy function name for compatibility
async def get_jwt_token(*args: Any, **kwargs: Any) -> str:
    """Legacy alias for get_jwt."""
    return await get_jwt(*args, **kwargs)


# Convenience functions for specific JWT types
async def get_jwt_database(
    workspace_group_id: Optional[str] = None,
    server_url: str = "https://authsvc.singlestore.com/auth/iam/database",
    provider: Optional[CloudProviderClient] = None,
    additional_params: Optional[dict[str, str]] = None,
    assume_role_identifier: Optional[str] = None,
    timeout: float = 10.0,
    logger: Optional[Logger] = None,
    **kwargs: Any,
) -> str:
    """
    Get a JWT for database access.

    Args:
        workspace_group_id: Workspace group ID (optional - can be None or empty string)
        server_url: Authentication server URL (defaults to production)
        provider: Optional provider client (will auto-detect if not provided)
        additional_params: Additional provider-specific parameters
        assume_role_identifier: Role to assume before getting JWT
        timeout: Request timeout in seconds
        logger: Optional logger instance
        **kwargs: Additional options

    Returns:
        str: JWT string for database access
    """
    return await get_jwt(
        jwt_type=JWTType.DATABASE_ACCESS,
        workspace_group_id=workspace_group_id,
        server_url=server_url,
        provider=provider,
        additional_params=additional_params,
        assume_role_identifier=assume_role_identifier,
        timeout=timeout,
        logger=logger,
        **kwargs,
    )


async def get_jwt_api(
    workspace_group_id: Optional[str] = None,
    server_url: str = "https://authsvc.singlestore.com/auth/iam/api",
    provider: Optional[CloudProviderClient] = None,
    additional_params: Optional[dict[str, str]] = None,
    assume_role_identifier: Optional[str] = None,
    timeout: float = 10.0,
    logger: Optional[Logger] = None,
    **kwargs: Any,
) -> str:
    """
    Get a JWT for API gateway access.

    Args:
        server_url: Authentication server URL (defaults to production)
        provider: Optional provider client (will auto-detect if not provided)
        additional_params: Additional provider-specific parameters
        assume_role_identifier: Role to assume before getting JWT
        timeout: Request timeout in seconds
        logger: Optional logger instance
        **kwargs: Additional options

    Returns:
        str: JWT string for API gateway access
    """
    return await get_jwt(
        jwt_type=JWTType.API_GATEWAY_ACCESS,
        workspace_group_id=workspace_group_id,
        server_url=server_url,
        provider=provider,
        additional_params=additional_params,
        assume_role_identifier=assume_role_identifier,
        timeout=timeout,
        logger=logger,
        **kwargs,
    )
